### DISCUSSION TITLE
Libraries to include under Windows VSIDE

### DISCUSSION THREAD

**Discussion by TryerGit** [google/or-tools#3903] [Build and configuration issues] [answered]
Hello,

Thanks to help from this forum, I have been able to install or-tools successfully on Windows, VS2022.

I am trying to follow the example given here: https://developers.google.com/optimization/introduction/cpp

However, I am doing this in Windows under Visual Studio IDE with the idea of stepping through the code. So, within the IDE, when I copy the file and provide C:\Program Files (x86)\ortools\include as the additional include directories, the code looks fine with no IDE generated red squiggly line underneath.

In Debug mode, I set up a breakpoint in the first line of `int main()`. On building it, there are a bunch of errors of the type:

Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	E0786	member "absl::lts_20230125::container_internal::internal_compressed_tuple::Storage<T, I, UseBase>::in_place_t [with T=size_t, I=0Ui64, UseBase=false]" is not a valid class member template	first_ortools_project	C:\Program Files (x86)\ortools\include\absl\container\internal\compressed_tuple.h	89	

Obviously, I seem to be missing some libraries to include in the Linker part of the IDE. Could it please be pointed out what these would be? I also provided C:\Program Files (x86)\ortools\lib as additional library directories under the Linker option of VSIDE. Also, in additional dependencies, I have specified ortools_full.lib

The tutorial getting started link above seems to assume that one is operating under a gnu system where it requires the user to use `make`

Is it possible to set it up so that one can step through the code of the C++ examples that go with the documentation?

I did do a search on SO as well as previous issues here under "Visual Studio Setup" but they do not seem to be about how to setup a Visual Studio project and step through the code.

Thank you.

**Comment by lperron** [comment]
You should look at https://github.com/or-tools/cmake_or-tools and import the cmake in your environment

**Reply by TryerGit** [reply]
Thank you. I downloaded the zipped file from that github page, extracted it into a folder and opened the folder containing the outermost CMakeLists.txt file in Visual Studio. Then, I opened up the developer power shell terminal within Visual Studio. Then, I issued the two commands specified on that page:

cmake -S. -Bbuild
cmake --build build -v

This then took a good half hour to complete and ended up creating a build/_deps folder which is about 5.7 GB in size. 

While I am not a complete novice to building using .vcxproj files in Windows and makefiles in Linux, and have used CPLEX and boost libraries in C++ projects, amongst others, Cmake is completely new to me. I am astounded that a simple hello world program should involve so much work in the background. Please pin this to my ignorance about CMake and Or-tools and how they work. 

My question at this stage is, would it take so long to build and use every or-tools project I embark upon, or is this a one time initial setup cost?

Thank you for your patience and time.

**Reply by ozan-cb** [reply]
You can use the `--parallel <num_cores>` option in the build step to speed up your build.

This is a one time cost if you are not modifying the source code of OR-Tools and are only using it as an external library in your CMake projects. See [Consuming OR-Tools in a CMake project](https://github.com/google/or-tools/blob/stable/cmake/README.md#consuming-or-tools-in-a-cmake-project).

**Reply by TryerGit** [reply]
Thank you. In hope of better luck, I reinstalled or tools by following the instructions provided here:

https://developers.google.com/optimization/install/cpp/source_linux

, this time, on a linux machine. Everything worked fine and I installed it by following the two commands:

cmake --build build --config Release --target all -j -v

and

sudo cmake --build build --config Release --target install -v

The last command, according to the documentation, should install OR-Tools "on my operating system" (I had to do sudo because in the final step, something was needed to be copied to /usr/bin, if I remember correctly). By operating system, I imagine in the appropriate place so that the code examples provided officially by this repository can be easily built and run without having the user to modify library paths/include paths.

Then, on VSCode, I installed CMake Tools extension -- the official extension from Microsoft. 

Now, I navigated to the folder where I extracted the zipped file from the repository:  https://github.com/or-tools/cmake_or-tools. Then, as suggested by the link "Consuming OR-Tools in a CMake project", I modified the CMakeList.txt found in the outermost folder of the folder by adding two lines thus (the added lines are indicated by C style comments):

```
cmake_minimum_required(VERSION 3.14.7)

if(POLICY CMP0022)
  cmake_policy(SET CMP0022 NEW)
endif()

# option() honors normal variables.
if(POLICY CMP0077)
  cmake_policy(SET CMP0077 NEW)
endif()

# Set default build type to Release if empty.
set(CMAKE_BUILD_TYPE "Release" CACHE STRING
  "Choose the type of build, options are: Debug Release RelWithDebInfo MinSizeRel.")

project(CMakeORTools VERSION 1.0 LANGUAGES CXX)

find_package(ortools CONFIG REQUIRED) // Added new
target_link_libraries(FooApp ortools::ortools) // Added new

include(CTest)
include(GNUInstallDirs)

if(UNIX)
  option(BUILD_SHARED_LIBS "Build shared libraries(.so)." ON)
else()
  # only support for static build on Windows.
  set(BUILD_SHARED_LIBS OFF)
endif()

if(BUILD_SHARED_LIBS)
  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
endif()

include(FetchContent)
FetchContent_Declare(
  ortools
  GIT_REPOSITORY https://github.com/google/or-tools.git
  GIT_TAG        main
)
set(BUILD_DEPS ON)
set(BUILD_DOC OFF)
set(BUILD_SAMPLES OFF)
set(BUILD_EXAMPLES OFF)
FetchContent_MakeAvailable(ortools)

add_subdirectory(Foo)
add_subdirectory(FooApp)

# Install
install(EXPORT ${PROJECT_NAME}Targets
  NAMESPACE ${PROJECT_NAME}::
  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
  COMPONENT Devel)
include(CMakePackageConfigHelpers)
configure_package_config_file(cmake/${PROJECT_NAME}Config.cmake.in
  "${PROJECT_BINARY_DIR}/${PROJECT_NAME}Config.cmake"
  INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}"
  NO_SET_AND_CHECK_MACRO
  NO_CHECK_REQUIRED_COMPONENTS_MACRO)
write_basic_package_version_file(
  "${PROJECT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
  COMPATIBILITY SameMajorVersion)
install(
  FILES
  "${PROJECT_BINARY_DIR}/${PROJECT_NAME}Config.cmake"
  "${PROJECT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
  DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}"
  COMPONENT Devel)

```

Now, VSCode CMake Extension reports the following error:

```
CMake Error at CMakeLists.txt:19 (target_link_libraries):Cannot specify link libraries for target "FooApp" which is not built by
this project.
```

Should I specify a different name than FooApp?

Thank you.

**Reply by ozan-cb** [reply, marked as answer]
After installing OR-Tools on your system as you have already did, in a clean directory with a CMakeLists.txt as:
```
cmake_minimum_required(VERSION 3.14)
project(myproj VERSION 1.0)

find_package(ortools CONFIG REQUIRED)

add_executable(myapp main.cpp)
target_link_libraries(myapp ortools::ortools)
```
and a `main.cpp` file containing some OR-Tools use should work. Also as noted in the documentation, you might need to supply a [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure).

It seems like the `cmake_or-tools` example project in your message fetches OR-Tools from Github and builds it from scratch, that is what you were trying to avoid when you were installing OR-Tools on your system. If fetching and building OR-Tools for your every new project that is acceptable, you can still use that method but then you should put your
`target_link_libraries(FooApp ortools::ortools) // Added new`
after this line:
`add_subdirectory(FooApp)`
so that CMake knows that FooApp exists.

**Comment by TryerGit** [comment]
@ozan-cb  Thank you. Your suggestion worked. I had to do a bit of modification since I was already linking to CPLEX and Boost and wanted to get cmake to work with those in addition to OR-Tools.
So, gave the following command: 

target_link_libraries(CMakeProject cplex-library Boost::boost ortools::ortools) since I believe CMake has to accept all target link libraries using a single command in a single line and was able to verify that it works with all 3 libraries.

**Comment by TryerGit** [comment]
Sorry. I have to reopen discussion. I ended up having to reinstall. But that went without any errors. I know that the install went well because I am able to run the following without any errors:

```
C:\Program Files (x86)\ortools\bin>assignment_groups_mip
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1694008714.640592   16440 assignment_groups_mip.cc:218] Total cost = 239


I0000 00:00:1694008714.641816   16440 assignment_groups_mip.cc:224] Worker 0 assigned to task 4.  Cost: 50
I0000 00:00:1694008714.642097   16440 assignment_groups_mip.cc:224] Worker 1 assigned to task 2.  Cost: 55
I0000 00:00:1694008714.642364   16440 assignment_groups_mip.cc:224] Worker 5 assigned to task 5.  Cost: 31
I0000 00:00:1694008714.642631   16440 assignment_groups_mip.cc:224] Worker 6 assigned to task 3.  Cost: 41
I0000 00:00:1694008714.642877   16440 assignment_groups_mip.cc:224] Worker 10 assigned to task 0.  Cost: 17
I0000 00:00:1694008714.643144   16440 assignment_groups_mip.cc:224] Worker 11 assigned to task 1.  Cost: 45

C:\Program Files (x86)\ortools\bin>
```

I presume the above indicates that OR Tools has installed properly.

Now, via code, I have a very minimal `CML.txt` thus:

```
cmake_minimum_required(VERSION 3.8)
project ("CMakeProject" LANGUAGES C CXX)
add_executable (CMakeProject code/Main.cpp)
message(STATUS "System: ${CMAKE_SYSTEM}")
message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
message(STATUS "Compiler version: ${CMAKE_CXX_COMPILER_VERSION}")
find_package(ortools CONFIG REQUIRED)
target_link_libraries(CMakeProject ortools::ortools)
```

On opening this in VSIDE 2022 with a very simple OrTools C++ program, I obtain the following error:

```
1> [CMake] -- System: Windows-10.0.22621
1> [CMake] -- Build type: Debug
1> [CMake] -- Compiler version: 19.37.32822.0
1> [CMake] CMake Error at C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
1> [CMake]   Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
1> [CMake] Call Stack (most recent call first):
1> [CMake]   C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
1> [CMake]   C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.26/Modules/FindZLIB.cmake:200 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
1> [CMake]   C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.26/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
1> [CMake]   C:/Program Files (x86)/ortools/lib/cmake/ortools/ortoolsConfig.cmake:25 (find_dependency)
1> [CMake]   CMakeLists.txt:8 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
```

Is it possible to figure out what is going wrong here?

Thanks.

**Comment by TryerGit** [comment]
To add to this -- before `find_package()`, I added `set(CMAKE_PREFIX_PATH "C:\\Program Files (x86)\\ortools")`

This ends up *compiling* fine without any CMake errors, yet, build/linking gives rise to a bunch of linking errors like so:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in Main.cpp.obj	D:\trials\cmake\windows\build\x64-Debug\trials	D:\trials\cmake\windows\build\x64-Debug\ortools.lib(linear_solver.obj)	1
```

**Comment by lperron** [comment]
You are mixing debug and release libraries.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by TryerGit** [reply]
Yes indeed. This fixes it. I had to create a new release configuration in VSIDE.

**Comment by TryerGit** [comment]
I have a follow up question. 

After having setup my IDE to work in Release mode, I am able to run ORTools, however, I am not able to setup break points and step through the code in debug mode though.

Is the fact that I am unable to run in Debug mode in Visual Studio on Windows (when I link in Debug mode, I get a bunch of errors which you pointed out above) a result of having run 

`cmake --build build --config Release --target INSTALL -v`

in the final installation step of the document https://developers.google.com/optimization/install/cpp/source_windows?

Had I run `cmake --build build --config Debug --target INSTALL -v`

would I be able to step through the code in Debug mode? If yes, then, that is quite suboptimal because I want to have the ability to run in Debug mode as well as Release mode without having to build ortools again and the IDE provides for choosing the appropriate configuration to run under.

Surprisingly, in Linux, even though I ended up giving the same final command as in Windows above, I am able to setup break points and step through the code in Debug mode, hence the question. This leads me to believe that it is something else is the cause of my not being able to run in Debug mode in Windows VSIDE.

In VS IDE, in Project -> CMake Settings menu item -> CMake Variables and Cache -> show advanced variables, I am able to see that ZLIB_LIBRARY_DEBUG has value ZLIB_LIBRARY_DEBUG-NOTFOUND, while ZLIB_LIBRARY_RELEASE has value C:/Program Files (x86)/ortools/lib/zlib.lib

Thank you. I am close to fully understanding the configuration needed to run ortools on linux and windows but not yet sufficiently there as yet and hence these persistent questions!
